app/services/lara_serialization_helper.rb

```diff
@@ -4,50 +4,66 @@ def initialize
     @entries = {}
   end
 
-  def cache_item(ref_id, item)
-    @entries[ref_id] = item
+  def export(item)
+    results = item.export
+    results[:type] = item.class.name
+    results[:ref_id] = key(item)
+    if item.respond_to?(:interactive) && item.interactive
+      results[:interactive_ref_id] = key(item.interactive)
+    end
+    if item.respond_to?(:embeddable) && item.embeddable
+      results[:embeddable_ref_id] = key(item.embeddable)
+    end
+    if item.respond_to?(:linked_interactive) && item.linked_interactive
+      # Why is linked interactive fully serialized while other linked items use reference only?
+      # Not sure. It might be related to the fact, that linked interactives can be linked across multiple pages
+      # or even activities. However, it doesn't seem it would work if we export and import activity with this
+      # interactive missing. I guess it could be double checked and fixed in the future.
+      results[:linked_interactive] = self.export(item.linked_interactive)
+    end
+    results
   end
 
-  def cache_item_copy(original_item, copy_of_item)
-    cache_item(key(original_item), copy_of_item)
-  end
+  def import(item_hash)
+    existing_item = lookup_item(item_hash[:ref_id])
+    return existing_item if existing_item
 
-  def lookup_item(key_sting)
-    @entries[key_sting]
+    item = item_hash[:type].constantize.import(item_hash.except(:type, :ref_id, :interactive_ref_id, :embeddable_ref_id, :linked_interactive))
+    item.save!(validate: false)
+    unless item_hash[:ref_id]
+      # This is only for backward compatibility when not all the embeddables were defining ref_id in export hash.
+      item_hash[:ref_id] = key(item)
+    end
+    cache_item(item_hash[:ref_id], item)
+    item
   end
 
-  def lookup_new_item(original_item)
-    lookup_item(key(original_item))
+  def set_references(item_hash)
+    item = lookup_item(item_hash[:ref_id])
+    if item_hash[:interactive_ref_id] && item.respond_to?(:interactive=)
+      item.interactive = lookup_item(item_hash[:interactive_ref_id])
+    end
+    if item_hash[:embeddable_ref_id] && item.respond_to?(:embeddable=)
+      item.embeddable = lookup_item(item_hash[:embeddable_ref_id])
+    end
+    if item_hash[:linked_interactive] && item.respond_to?(:linked_interactive)
+      item.linked_interactive = import(item_hash[:linked_interactive])
```

Comment:
As an exercise I mentally checked if this will work with a forward reference. This also helped me figure out what we'd need to do so that linked_interactive items don't need to be fully serialized.

Let's say the structure is:
```
Page 1: { interactive A: {linked_interactive: interactive B}}
Page 2: { interactive B: {linked_interactive: interactive C}}
Page 3: { interactive C }
```

1. interactive A on page 1 will be imported. 
2. set_references will be called with interactive A's hash. 
3. This will trigger an import of the fully serialized interactive B.  
4. This import of the interactive B will not find it in `@entries`, so it will create a new object and set the ref_id if one is not set
5. Interactive B will be cached with the ref_id.
6. LightweightActivity#import will then call import on Page 2
7. Interactive B will be imported again. 
8. The same ref_id should be defined in both places, so this second import will shortcut at line 29. 
9. Later in InteractivePage#import set_references will be called on Interactive B's hash. 
10. Interactive C will now be imported. 
(the rest of the steps are not necessary to trace)

If we were not fully serializing the interactives, then on step 3 Interactive B would not be imported. Instead lookup would be called and it would not find anything.

To make this system work for forward references and not require fully serializing the interactives. The resolving of references in set_references needs to happen after the top level item has been imported. In the first pass all of the items will be created without references. Then in the second pass all of these references are resolved.  

This can be implemented by creating an array of reference_objects during the first pass.  This reference object would look like 
```
{ item: [item that is referring to a reference], 
  setter: [name of the setter to set the reference for example 'linked_interactive=' ], 
  ref_id: [id of the reference]} 
```
So in the second pass each of these reference objects would be processed with:
`ref_obj.item.send(ref_obj.setter, lookup_item(ref_obj.ref_id))`

This second pass could be in a `resolve_references` method on LARASerializationHelper. 
And the import methods of activity and page would need another option so they won't call helper.resolve_references if it is going to be called by the parent.

Reply:
> To make this system work for forward references and not require fully serializing the interactives. 

Yes, that's definitely possible, I thought about it, and your approach is one of the possible solutions for sure. The main reason I didn't do it now was to make this changeset a bit smaller and also to stay backward compatible. However, I agree it'd be great to fix it, and since we started this discussion, I guess it's the right moment.

Your approach sounds fine to me. When I was working on that, I thought about something slightly different:

"The first pass" would happen in the top-level object that is being imported, before we process pages. So, in practice, it'd be a sequence or an activity. Same as now, the first pass would ignore all the references, it would just import/create items. And that's pretty much all. Now, we can do "the second pass" aka `set_references` in the same place as we do now (page import code). We would be able to use references there, no matter if it's forward or backward reference, as we did initial import in the parent.

The perfect place to do to this initial, first pass would be `LaraSerializationHelper` constructor. Note that there's already logic that initializes this helper in the right place (sequence, activity or page). This will let us avoid problems / new options like:

> And the import methods of activity and page would need another option so they won't call helper.resolve_references if it is going to be called by the parent.

All the other code will stay the same, as `LaraSerializationHelper` already uses cache if possible.

Let me know what you think about it, or which version you prefer. I described my initial ideas, but not sure if I covered all the cases. Also, should we care about being backward compatible? I guess it'd be possible just by leaving existing code. New one will use `linked_interactive_ref_id` key instead of `linked_interactive`.

In either way, the additional benefit (besides making things consistent) will be that embeddable and interactive references will work accross different pages (if we ever need that). Right now, only linked_interactives can do it.

Reply:
However after going through all of this, I'm going to merge this PR in as is, so Trudi and Sarah can start testing. And so this PR doesn't get any bigger.

Also I think it is better to postpone working on this more until the [interactive overlay story](https://www.pivotaltracker.com/story/show/161350185) and the [wrapped embeddable authoring story](https://www.pivotaltracker.com/story/show/161843670) are done.   Unfortunately that probably means you won't be able to work on this more. ☹️ But maybe things will go fast. Perhaps we'll be able to convince folks that taking care of the infinite loop/recursion issue makes it worth spending a bit more time on this.

If you do work on it more here are some more notes:

I think we should handle backward compatibility. The code for importing could check for linked_interactive and just pull out the ref_id from it and ignore the rest.

I think either approach is good. I guess my preference is putting the second pass (resolve references) at the top level. And also to limit the number of places that need to know about the reference field names.   In the PR's current code `embeddable_ref_id` is directly used in the code on the export, the first pass of the import, and the set_references (second pass).  Using the approach of reference objects we could eliminate of the use of it on the second pass.  

We could even eliminate hard coded field names from the import all together if the import code could determine a field was a reference just from the format.  For example it could look for `*_ref_id` fields and automatically assume that was a reference. However, that seems a little fragile. In systems that do this I've usually seen them use a pattern like `field: { $ref_id: 123 }`. That is how the JSONPointer spec (used by jsonschema) does it: https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03